src/Spice86.Core/Emulator/Devices/Video/VideoMemory.cs

```diff
@@ -1,32 +1,260 @@
 namespace Spice86.Core.Emulator.Devices.Video;
 
-using Spice86.Core.Emulator.Memory;
-using Spice86.Shared.Interfaces;
+using Spice86.Core.Emulator.Devices.Video.Registers;
+using Spice86.Core.Emulator.Devices.Video.Registers.Graphics;
+
+using System.Diagnostics;
 
 /// <summary>
-/// A wrapper class for the video card that implements the IMemoryDevice interface.
+///     A wrapper class for the video card that implements the IMemoryDevice interface.
 /// </summary>
-public class VideoMemory : IMemoryDevice {
-    private readonly IVideoCard _videoCard;
-    private readonly uint _baseAddress;
+public class VideoMemory : IVideoMemory {
+    private readonly byte[] _latches;
+    private readonly IVideoState _state;
 
-    public VideoMemory(uint size, IVideoCard videoCard, uint baseAddress) {
-        _videoCard = videoCard;
-        _baseAddress = baseAddress;
-        Size = size;
-    }
-    
-    public uint Size {
-        get;
+    /// <summary>
+    ///     Create a new instance of the video memory.
+    /// </summary>
+    /// <param name="state"></param>
+    public VideoMemory(IVideoState state) {
+        _state = state;
+        Planes = new byte[4, 0x20000];
+        _latches = new byte[4];
+        Size = 0x20000;
     }
+
+    /// <inheritdoc />
+    public uint Size { get; }
+
+    /// <inheritdoc />
     public byte Read(uint address) {
```

Comment:
If I understand well this is done for each pixel displayed.
I am wondering if it would make sense to have a cache for this:
 - first read would populate the cache
 - second read would read from the cache
 writing at the planes / registers / latches would invalidate the cache entries.

How is performance compared to existing solution?

Reply:
This is for reading video memory from the application side, not the card/renderer side.

This is mostly used for reading exisiting graphics, for instance when moving a mouse pointer over them. So not used a often as writing to video memory.

Performance is quite good actually. Not worth building a complicated caching mechanism for in my opinion.